src/prognostic_equations/implicit/approx_jacobian.jl

```diff
@@ -0,0 +1,879 @@
+using ClimaCore.MatrixFields
+import LinearAlgebra: I, Adjoint
+
+abstract type DerivativeFlag end
+struct UseDerivative <: DerivativeFlag end
+struct IgnoreDerivative <: DerivativeFlag end
+use_derivative(::UseDerivative) = true
+use_derivative(::IgnoreDerivative) = false
+
+"""
+    ApproxJacobian(;
+        [diffusion_flag],
+        [topography_flag],
+        [sgs_advection_flag],
+        [approximate_solve_iters],
+    )
+
+An approximation of the exact `ImplicitEquationJacobian`, which is updated using
+analytically derived tendency derivatives and inverted using a specialized
+nested linear solver. The approximation also has flags for setting certain
+derivatives to 0; these are represented as `DerivativeFlag`s (either
+`UseDerivative` or `Ignoreerivative`) instead of `Bool`s for type-stability.
+
+# Keyword Arguments
+
+- `topography_flag::DerivativeFlag`: whether the derivative of vertical
+  contravariant velocity with respect to horizontal covariant velocity should be
+  computed or approximated as 0
+- `diffusion_flag::DerivativeFlag`: whether the derivatives of the grid-scale
+  diffusion tendency should be computed or approximated as 0
+- `sgs_advection_flag::DerivativeFlag`: whether the derivatives of the
+  subgrid-scale advection tendency should be computed or approximated as 0
+- `sgs_mass_flux_flag::DerivativeFlag`: whether the derivatives of the
+  subgrid-scale mass flux tendency should be computed or approximated as 0
+- `approximate_solve_iters::Int`: number of iterations to take for the
+  approximate linear solve required when the `diffusion_flag` is `UseDerivative`
+"""
+@kwdef struct ApproxJacobian{F1, F2, F3, F4} <: JacobianAlgorithm
+    topography_flag::F1 = nothing
+    diffusion_flag::F2 = nothing
+    sgs_advection_flag::F3 = nothing
+    sgs_mass_flux_flag::F4 = nothing
+    approximate_solve_iters::Int = 1
+end
+
+function jacobian_cache(alg::ApproxJacobian, Y, atmos)
+    topography_flag = if isnothing(alg.topography_flag)
+        has_topography(axes(Y.c)) ? UseDerivative() : IgnoreDerivative()
+    else
+        alg.topography_flag
+    end
+    diffusion_flag = if isnothing(alg.diffusion_flag)
+        atmos.diff_mode == Implicit() ? UseDerivative() : IgnoreDerivative()
+    else
+        alg.diffusion_flag
+    end
+    sgs_advection_flag = if isnothing(alg.sgs_advection_flag)
+        atmos.sgs_adv_mode == Implicit() ? UseDerivative() : IgnoreDerivative()
+    else
+        alg.sgs_advection_flag
+    end
+    sgs_mass_flux_flag = if isnothing(alg.sgs_mass_flux_flag)
+        atmos.sgs_adv_mode == Implicit() ? UseDerivative() : IgnoreDerivative()
+    else
+        alg.sgs_mass_flux_flag
+    end
```

Comment:
```suggestion
    treat_∂(::Nothing, mode::Bool) = mode ? UseDerivative() : IgnoreDerivative()
    treat_∂(::Nothing, ::Implicit) = UseDerivative()
    treat_∂(::Nothing, ::Explicit) = IgnoreDerivative()
    treat_∂(flag, _) = flag

    topography_flag = treat_∂(alg.topography_flag, has_topography(axes(Y.c)))
    diffusion_flag = treat_∂(alg.diffusion_flag, atmos.diff_mode)
    sgs_advection_flag = treat_∂(alg.sgs_advection_flag, atmos.sgs_adv_mode)
    sgs_mass_flux_flag = treat_∂(alg.sgs_mass_flux_flag, atmos.sgs_adv_mode)
```
We could write a couple helpers here, if you think it's worth it.